course7/homeworks/patrickkostjens/complex/main.cc

```diff
@@ -3,27 +3,31 @@
 #include <boost/filesystem.hpp>
 #include <string>
 #include <dlfcn.h>
+#include <vector>
 
 namespace fs = boost::filesystem;
 
 int main(int argc, char** argv)
 {
 	Calculator c;
-	int plugin_count = 0;
-	for (const auto& it : fs::directory_iterator(fs::path(".")))
-		if (it.path().extension() == ".so" || it.path().extension() == ".dylib")
-			plugin_count++;
-
-	plugin *plugins = new plugin[plugin_count];
-	plugin_count = 0;
 	for (const auto& it : fs::directory_iterator(fs::path(".")))
 	{
 		if (it.path().extension() == ".so" || it.path().extension() == ".dylib")
 		{
-			void* module = dlopen(it.path().filename().c_str(), RTLD_NOW);
-			plugins[plugin_count] = reinterpret_cast<plugin>(dlsym(module, "execute"));
-			c.add_plugin(it.path().stem().string(), &plugins[plugin_count]);
-			plugin_count++;
+			std::string name = "./" + it.path().filename().string();
+			void* module = ::dlopen(name.c_str(), RTLD_NOW);
+			if (!module)
+			{
+				std::cerr << "Error loading module " << name << " : " << ::dlerror() << std::endl;
+				return 1;
+			}
+			plugin p = reinterpret_cast<plugin>(::dlsym(module, "execute"));
+			if (!p)
+			{
+				std::cerr << "Symbol not found from file " << name << ", : " << ::dlerror() << std::endl;
+				return 1;
```

Comment:
I'm not sure these returns should be here; they make the errors fatal. I don't think a failure to load a plugin should be fatal.